vignettes/d3-to-shiny-communication.Rmd

```diff
@@ -0,0 +1,98 @@
+---
+title: "D3 to Shiny communication"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{d3-to-shiny-communication}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(eval = FALSE)
+```
+
+
+Having a D3 plot update based on a changing Shiny input is very common. But, what about having an action that happened inside the D3 plot trigger something in Shiny?  For example, we may want Shiny to know that our D3 bar plot was clicked on, and know which bar was clicked. This ability becomes necessary when features such as drill-down, have to be included in the Shiny app. This can be accomplished by adding a call to a Shiny JS function from inside the D3 script. And in the Shiny app, an event observer can be added that looks for the JS function call. 
+
+## Example
+
+In this example, the bar plot will pass to Shiny the data value of the bar plot that was clicked by the user.  Shiny will then display a notification containing the value of the bar that was clicked on.
+
+Here is the D3 script:
+
+```js
+// !preview r2d3 data=c(0.3, 0.6, 0.8, 0.95, 0.40, 0.20)
+
+var barHeight = Math.ceil(height / data.length);
+
+svg.selectAll('rect')
+  .data(data)
+  .enter().append('rect')
+    .attr('width', function(d) { return d * width; })
+    .attr('height', barHeight)
+    .attr('y', function(d, i) { return i * barHeight; })
+    .attr('fill', 'steelblue')
+    .attr("id", function(d) { return d; })
+    .on("click", function(){
+            Shiny.onInputChange(
+              "bar_clicked", 
+              d3.select(this).attr("id"));
```

Comment:
Let's call the made up attribute `d` rather than `id` (since the semantics of id are more about uniqueness than values)